File: tests/accesses/shift/accesses_shift.ml
open Optitrust
open Prelude

let _ = Flags.check_validity := true
let _ = Flags.recompute_resources_between_steps := true

let _ = Run.script_cpp (fun _ ->
  !! Accesses.shift_var ~factor:(trm_float 5.0) [nbMulti; cTopFunDef "test_var"; cVarDef "x"];
  !! Accesses.shift_var ~factor:(trm_int 1) [nbMulti; cTopFunDef "test_var_inv"; cVarDef "s"];

  (* TODO: !! Accesses.shift ~factor:(trm_float 5.0) [nbMulti; cTopFunDef "test_array"; sExpr "t[0]"];
  !! Accesses.shift ~factor:(trm_float 5.0) [nbMulti; cTopFunDef "main"; sExpr "t[i]"]; *)
)

File: tests/accesses/shift/accesses_shift.cpp
#include <optitrust.h>

void test_var() {
  __pure();

  double x = 0.0;
  x = 1.0;
  double y = x * 1.0;
  x = x * 2.0;
}

void test_var_inv(int* t, int n) {
  __reads("t ~> Matrix1(n)");

  int s = 0;
  for (int i = 0; i < n; i++) {
    __spreserves("&s ~> Cell");
    s = s + 1;
  }
}


void test_array() {
  double t[2] = { 1., 2 };
  t[0] = t[0] * 1.0;
}


int main() {
   double t[3] = { 1., 2., 3. };
   double v = 2.0;
   int i = 0;
   t[i] = t[i] * v;
   return 0;
}

